File: leetcode/Container With Most Water.js
// https://leetcode.com/problems/container-with-most-water/
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  const N = height.length
  let area = 0
  let l = 0
  let r = N - 1

  while (l < r) {
    area = Math.max(area, Math.min(height[l], height[r]) * (r - l))
    if (height[r] > height[l]) ++l
    else --r
  }

  return area
}

class Tester {
  constructor(fun) {
    this.fun = fun
    this.total = 0
    this.fail = 0
    this.time = 0
  }

  get pass() {
    return this.total - this.fail
  }

  test(expected, ...input) {
    const start = performance.now()
    const output = this.fun(...input)
    this.time += performance.now() - start

    const isCorrect = output === expected

    console.log('Test', this.total, '...', isCorrect)

    if (!isCorrect) {
      console.error('Output:', output)
      console.error('Expected:', expected)
      ++this.fail
    }

    console.log('')
    ++this.total
  }

  report() {
    console.log(`Passed ${this.pass} / ${this.total} tests
`)
    console.log(`Ran in ${this.time.toFixed(4)}ms`)
  }
}

const t = new Tester(maxArea)

t.test(49, [1, 8, 6, 2, 5, 4, 8, 3, 7])
t.test(1, [1, 1])
t.test(0, [0, 2])

t.report()

File: leetcode/Container With Most Water.py
# https://leetcode.com/problems/container-with-most-water/
from typing import List
from tester import Tester


class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        max_area = 0
        while l < r:
            max_area = max(max_area, min(height[l], height[r]) * (r - l))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return max_area


t = Tester(Solution())


t.test(49, [1, 8, 6, 2, 5, 4, 8, 3, 7])
t.test(1, [1, 1])
t.test(16, [4, 3, 2, 1, 4])
t.test(2, [1, 2, 1])

t.report()
